src/librbd/crypto/luks/FlattenRequest.h

```diff
@@ -0,0 +1,71 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_LIBRBD_CRYPTO_LUKS_FLATTEN_REQUEST_H
+#define CEPH_LIBRBD_CRYPTO_LUKS_FLATTEN_REQUEST_H
+
+#include "librbd/ImageCtx.h"
+
+namespace librbd {
+
+namespace crypto {
+namespace luks {
+
+template <typename I>
+class FlattenRequest {
+public:
+    using EncryptionFormat = decltype(I::encryption_format);
+
+    static FlattenRequest* create(I* image_ctx, Context* on_finish) {
+      return new FlattenRequest(image_ctx, on_finish);
+    }
+
+    FlattenRequest(I* image_ctx, Context* on_finish);
+    void send();
+
+private:
+  /**
+   * @verbatim
+   *
+   * <start>
+   *    |
+   *    v
+   * SHUTDOWN_CRYPTO
+   *    |
+   *    v
+   * READ_HEADER
+   *    |
+   *    v
+   * WRITE_HEADER (replacing magic back from RBDL to LUKS if needed)
+   *    |
+   *    v
+   * FLUSH
+   *    |
+   *    v
+   * <finish> (+ RESTORE_CRYPTO)
```

Comment:
I am not sure I might be missing something, but is it safe just to shut down the crypto temporarily? Don't we need some protection for concurrent i/o during this window? 

AFAIR our discussion long time ago you were going to open a separate image context just to be able to run it without the crypto layer. And I seem not to find that code in the current implementation?

Reply:
Well, we can add a flush, and relay on `exclusive_lock`, which is acquired when flattening an image (assuming image supports exclusive lock).
So maybe we can add a flush to protect against in-flight writes?

But for reads, I don't think there's any mechanism we can use?

I reverted the old method of using a new ImageCtx after this review:
https://github.com/ceph/ceph/pull/40363#discussion_r855573758

Reply:
Actually, I expect the problem only when `exclusive_lock` is used. I don't expect "concurent io" for the same image context, when a user runs flatten for an image without `exclusive_lock`. But when you run for an image with the exclusive lock enabled and the image is currently in use, the operation will be just proxied to the lock owner and you will have both flatten and user i/o there.

If we don't want to open a new image context for this then may be we can temporarily block i/o?

Reply:
> If we don't want to open a new image context for this then may be we can temporarily block i/o?

I don't see any way do that in the current code (correct me if I'm wrong).
We can add a way to do this (say, by adding `block_io` to `librbd::io::Dipspatcher`), but we will also need to add a flag that allows to skip the block (for the raw IO that is triggered from `luks::FlattenRequest` to read/write the LUKS header).

But I think it will be easier to just add `librbd::io::READ_FLAG_SKIP_CRYPTO`, and `librbd::io::WRITE_FLAG_SKIP_CRYPTO`, which also means adding a new `write_flags` to `ImageDispatchSpec::create_write`, and changing all existing layers to properly propagate it.

What do you think?

Reply:
We already have `block_writes` but no `block_reads`...

Anyway, your idea with `SKIP_CRYPTO` flags looks interesting to me.

Reply:
@idryomov your thoughts?

This will mean also adding `bool skip_crypto` to `CopyupRequest` constructor, and to `ObjectDispatcher::prepare_copyup`, as well as add `write_flags` to `ImageDispatchInterface::write`.

And perhaps other changes I have not yet thought of.

Reply:
> Anyway, your idea with `SKIP_CRYPTO` flags looks interesting to me.

I suggested `SKIP_CRYPTO` flags a while back but IIRC Or ran into a complication with negative image offsets arising for the area left of `m_data_offset` and eventually fell back to the temporary crypto shutdown approach.  I think the crypto override approach has the potential to be a cleaner, at least conceptually, so I'm for it.

Reply:
It seems ok to me. I am not sure about `SKIP_CRYPTO_AND_CACHE` flag name. The first my though was that we could have two: `SKIP_CRYPTO` and `SKIP_CACHE`. On the other hand, I suppose "skip crypto" assumes that we don't want to cache the result, so may be just name the flag `SKIP_CRYPTO`?

Also I see you renamed `OBJECT_WRITE_FLAG_CREATE_EXCLUSIVE` to `WRITE_FLAG_CREATE_EXCLUSIVE`. I think I understand the reason, on the other hand from the new name it is not very clear what we have to create exclusive. May be it should become something like `WRITE_FLAG_OBJECT_CREATE_EXCLUSIVE`?

Reply:
> It seems ok to me. I am not sure about `SKIP_CRYPTO_AND_CACHE` flag name. The first my though was that we could have two: `SKIP_CRYPTO` and `SKIP_CACHE`. On the other hand, I suppose "skip crypto" assumes that we don't want to cache the result, so may be just name the flag `SKIP_CRYPTO`?
> 

Well I don't bother adding "skip crypto" functionality to the cache layers. So splitting up "skip_crypto_and_cache" will be misleading, since you won't be actually able to skip crypto without skipping the cache as well.
I think `SKIP_CRYPTO_AND_CACHE` is a "safer" name since it is explicit to future developers that skip_crypto is bundled with skipping the cache.
Of course this is not critical IMO, so I'm good with renaming it to just `skip_crypto`, but then we would need to add good commenting in each place we skip the cache with skip_crypto (e.g. ObjectCacher, SimpleScheduler, ParentCache, RWL).

> Also I see you renamed `OBJECT_WRITE_FLAG_CREATE_EXCLUSIVE` to `WRITE_FLAG_CREATE_EXCLUSIVE`. I think I understand the reason, on the other hand from the new name it is not very clear what we have to create exclusive. May be it should become something like `WRITE_FLAG_OBJECT_CREATE_EXCLUSIVE`?

Sure.

Reply:
> I think SKIP_CRYPTO_AND_CACHE is a "safer" name since it is explicit to future developers that skip_crypto is bundled with skipping the cache.

Ok. Actually I don't mind. I think my main concern is that the flag name is very long. Note, in the current code you made it `OBJECT_COPY_REQUEST_SKIP_CRYPTO_AND_CACHE` but it should be even longer `OBJECT_COPY_REQUEST_FLAG_SKIP_CRYPTO_AND_CACHE`.

Reply:
Hrm, this PR is huge as it is without these mechanical changes...  I'm inclined to defer the move away from the temporary crypto shutdown to another PR and have the discussion there, just to finally land the basics to improve upon.  @trociny do you agree?  Do you see any other issues?

Reply:
I am fine with addressing this in a separate PR. I am a bit concerned though that we are going to merge a code that may lead to data corruption during flatten (to me chances to have invalid (unencrypted) read/write if flatten is run concurrently are rather high). May be merge it with flattening encrypted clones disabled (at least in CLI) for now?

Reply:
May be merge it with flattening encrypted clones disabled (at least in CLI) for now?

+1 and not just in the CLI.  Ordinarily I would insist on such things being addressed in the same changeset but this PR is getting harder and harder to review with rebases, Github collapsing comments left and right, etc...

Reply:
> +1 and not just in the CLI.

OTOH this would break pretty much all encryption tests -- `luks-encryption.sh` and `TestCloneEncryption` in particular.  Given that Or is already working on this issue and it is supposed to be addressed ASAP, a test breaking check might not be worth it.  This is `main` branch in the middle of the development cycle after all...

Reply:
So I opened up a new PR for adding "skip_crypto" (#47601).
However, I've been re-thinking this after I discovered (using the help of @trociny) that `flatten` can be delegated, and I'm not sure this is the right way to go:

1. I think skip_crypto still does not fix all our problems. What if the lock owner who is delegated to perform the flatten does not have crypto loaded? I think we need to extend `librbd::watch_notify::FlattenPayload` with `bool with_encryption`, and if the lock owner does not have encryption loaded it should return `-EOPNOTSUPP`. In that case, the original caller should `acquire_lock` instead of `try_acquire_lock`.
2. The skip_crypto PR complicates many of the IO path components by adding yet another parameters, which frankly is quite is esoteric (it is only to be used in a very rare scenario of flattening a formatted clone, and useful only when the flatten operation is delegated). Perhaps we can try to make it more general by trying to replace the many instances of `bool skip_crypto_and_cache` with something like `int io_path_flags`, but the problem is I don't think we can define general propagations between `read_flags`, `write_flags`, `object_copy_flags` and the new `io_path_flags`. So we still have to parse down each bit of the flags and create a custom propagation in each point along the IO-path.
3. Given point (2) above, I'm not sure making many changes to the IO path is the right way. These changes are required strictly for the last phase of the flattening, which is changing the LUKS header, which involves just a simple read and write, which are not to be conflicted with any other IO (BTW the `skip_crypto_and_cache` flag also implicitly assumes that). I think this read-and-write operation should be done using a fresh image context. We could do it sort of externally, using `rbd_read` and `rbd_write` from the CLI, but the problem is that `flatten` detaches the child, and we may be left with a corrupted state if the read-and-write (which will be executed only after flatten completes) operation fails. So this leads me to think that the best way is to have this read-and-write operation as part of the flatten operation itself, as the final phase. This is what @trociny suggested and I implemented in the first place, but it was objected to because of possible edge cases. Perhaps we can re-think this and address any edge-cases? I think that from a code stability perspective, the blast-radius of making many changes to the IO-path is much larger.

Reply:
For (1), it is probably not a very good idea to force acquire the lock if the current lock owner does not supports encryption. The lock may be "stolen" again by the previous lock owner. And if we block steeling the lock during flatten, that would mean that if another client is doing write it will block until the flatten is complete.

On the other hand, I can hardly imagine a use case when someone uses an encrypted image without encryption. I don't think it will be a common case. So, may be just fail if the current lock owner does not support encryption?

And instead of adding `with_encryption` param to `Flatten` op, it might be easier to add a new `EncryptedFlatten` op? Then older clients (the lock owners) will automatically return `-EOPNOTSUPP` for an unknown operation.

As for (2,3). May be we could rewrite the LUKS header with just rados read/write operations (i.e. not involving the image context)? It is surely no problem to read the parent object, as it is not supposed to be changed. For writing, I also don't expect it would be problematic, because we will rewrite a region of the object that is not supposed to be modified by the "normal" io. So the only complication seems to manually calculate (using the image stripping format) object(s) and the region(s) to read/write the header.

Reply:
> As for (2,3). May be we could rewrite the LUKS header with just rados read/write operations (i.e. not involving the image context)? It is surely no problem to read the parent object, as it is not supposed to be changed. For writing, I also don't expect it would be problematic, because we will rewrite a region of the object that is not supposed to be modified by the "normal" io. So the only complication seems to manually calculate (using the image stripping format) object(s) and the region(s) to read/write the header.

I think this should not be too complicated.
We should copy code from `ImageReadRequest<I>::send_request`, and replace `ObjectDispatchSpec::create_read` with code copied from `ObjectReadRequest<I>::read_object()`.
On `-ENOENT` we should repeat the process on the parent image context.
I think that's it?

@idryomov your thoughts on this?

BTW this is not the first case where code in librbd directly uses librados, instead of using the rbd dispatches stack (i.e. `CopyupRequest<I>::copyup()`).

Reply:
Oh, I forgot we need a write as well.
This is more complex because there's updating the object map, which involves more code copying, and more complex IMO.
This seems fragile to me.
And not sure if there are no other bits of functionality that we may have missed, or may be added in the future to `ObjectRequest.cc`, and may leave the crypto flatten code broken.

Reply:
Perhaps just copy the bits from `ImageRequest.cc` to translate the header reads/writes to object requests, then use `ObjectDispatchSpec::create_read`, starting from the `CORE` (e.g. `ObjectRequest.cc`) layer, and disabling both copyup and "reading from parent", using appropriate `read_flags` / `write_flags` ?

I do see a little caveat in `ObjectRequest<I>::compute_parent_extents`, which calls `io::util::extent_to_file` (which is crypto aware due to its call to `remap_extents`), but perhaps we can overcome this using a new flag.

Reply:
What I had in mind was what we had had in older (octopus or earlier) versions of deep_copy/ObjectCopyRequest.cc, before it was changed to use `ObjectDispatchSpec::read` and just had used `m_src_io_ctx.aio_operate`. Though it should be much simpler in the current case as we don't need to deal with snapshots and the object map. I don't see why we would need to update the object_map. The object should already exist and we just rewriting the magic.

So the procedure should be:
1) calculate read object extent(s)
2) rados read the object(s) (if `-ENOENT` is returned restart with the parent)
3) calculate write object extent(s)
4) rados write the object(s).

It may still look complicated for such a simple operation as rewriting the header, and I don't insist we should go this way,though it still looks simpler than adding the skip flag.

Reply:
Hm. Not existing header would mean that the clone was not formatted yet. Would the flatten (not mentioning the image access) work as expected for this case?

Reply:
Consider this case:
Plaintext image a.
Image b is a clone of a, and formatted -> its LUKS header is on b's data, with magic RBDL (instead of LUKS).
Image c is a clone of b (not formatted) -> its LUKS header is on b's data, with magic RBDL.
User flattens image c.

Reply:
Wouldn't the object be marked as exist in the object_map after flatten is complete (before rewriting the header)?

Reply:
Because if it does not exit there might be a flow. Suppose, the flatten was interrupted and then the user run rm. It would skip removing the header object.

Reply:
Nope, due to a fix we put in this PR to have `operation::FlattenRequest` start flattening from the first object that corresponds to the encrypted data (as opposed to start from object #0).
Having flatten start from object #0 (where the crypto header starts) will cause negative offsets coming back from `extent_to_file`.

See this:
https://github.com/ceph/ceph/blob/f60190c668b14475e3d48c1adecb80c0013b4c37/src/librbd/operation/FlattenRequest.cc#L103

Reply:
> Because if it does not exit there might be a flow. Suppose, the flatten was interrupted and then the user run rm. It would skip removing the header object.

Actually, never mind. If the header was not copied there is nothing to remove. So there might be no issues here, though I would investigate different possibilities here.

Probably we need to have two "header flatten" instead of one at the end? The first "pre flatten" would make sure the header exists in the child before we start flattening.

Reply:
Thinking a little more. "pre flatten" seems not useful here. And indeed, it looks like the "header flatten" needs to update the object map.

Reply:
This all comes down to the implicit "unformatted clone" use case and I'm beginning to wonder if it is actually worth it in its current form.  What if we actually scrap it and instead add an explicit "same formatting as parent" format?  I think along the lines of the following:

- change `rbd_encryption_load` to require that clone is formatted and fail otherwise (e.g. by passing `DISABLE_READ_FROM_PARENT` in `LoadRequest::read`)
- introduce a new `RBD_ENCRYPTION_FORMAT_PARENT` format constant applicable only for formatting, not loading (same as `RBD_ENCRYPTION_FORMAT_LUKS{1,2}`)
- make `rbd_encryption_format` handle `RBD_ENCRYPTION_FORMAT_PARENT` format by triggering copyups for objects in the header area or by explicitly reading and writing them (in both cases the crypto layer wouldn't get in the way as the encryption can't possibly be loaded at that point)
- users would need to run `rbd encryption format <image-spec> parent` after `rbd clone` to get what is referred to as "unformatted clone" today
- no special handing in `rbd_flatten` other than RBDL -> LUKS magic fixup would be needed

Thougths?

Reply:
> * make `rbd_encryption_format` handle `RBD_ENCRYPTION_FORMAT_PARENT` format by triggering copyups for objects in the header area or by explicitly reading and writing them (in both cases the crypto layer wouldn't get in the way as the encryption can't possibly be loaded at that point)

How would you know the crypto header size?
Do you suggest to add code that enumerates all possible formats (which currently is only `LUKSEncryptionFormat`), and call a new function 'EncryptionFormat::get_header_size`, which will involve a new class `crypto::GetHeaderSizeRequest`, or possibly add a parameter `bool only_parse_header` to `crypto::LoadRequest`.

Reply:
@idryomov got an ever better suggestion from @dannyharnik :
Instead of adding new API to specifically copyup the parent LUKS header, and disallowing load before that,
how about triggering the copy-up of the the header right in `crypto::LoadRequest`, just before loading the crypto dispatch layers?

Reply:
> How would you know the crypto header size? Do you suggest to add code that enumerates all possible formats (which currently is only `LUKSEncryptionFormat`), and call a new function `EncryptionFormat::get_header_size`, which will involve a new class `crypto::GetHeaderSizeRequest`, or possibly add a parameter `bool only_parse_header` to `crypto::LoadRequest`.

The former -- only I think it should be a bit more general than `get_header_size` and return e.g. block size as well.  Then it could also serve as a building block for `rbd_encryption_info` which came up before in review of rbd-nbd changes (blksize nit):

> Perhaps a new rbd_encryption_info API that would return the format and the opts structure populated with that format's specifics, except for the passphrase of course ;) It should return the same results both with encryption loaded (just a quick dump from memory) and without it (go to RADOS).
>
> A new rbd encryption info CLI could be build on top of it, utilizing the "go to RADOS" mode. I see it being particularly useful for formatted clones, where the regular LUKS tools stop working.

Reply:
> @idryomov got an ever better suggestion from @dannyharnik : Instead of adding new API to specifically copyup the parent LUKS header, and disallowing load before that, how about triggering the copy-up of the the header right in `crypto::LoadRequest`, just before loading the crypto dispatch layers?

I don't think it is a good idea.  Conceptually, `rbd_encryption_load` is a read operation and as such should work with read-only cephx caps.

Reply:
> Thougths?

For the record, I must say I don't like it so much, for 2 reasons:
1. The encryption feature we have in librbd is purely data-centric. At least today, a user expects that if the image data contains a valid LUKS format, then `rbd_encryption_load` should succeed.
2. I believe the right "time" to copy the LUKS header is at "flatten" time, not at "clone" time. Basically you're wasting child space (although only a few MBs, but still a waste), because of a very premature copyup (which may never be needed, if that image will never be flattened), that is strictly there because of implementation considerations.

But anyhow, I don't see this debate as critical. So as far as I'm concerned I'm planning to go-on next week with implementing along the lines suggested above.

Reply:
+1 for what @orozery said. I do not object but I am not very happy with this solution. It would be nice if we could avoid this.

It could be easy if we cared only about the "cli" way to flatten an image. I.e. the `rbd flatten` could do `rbd_encryption_format` as the first step, before loading encryption.

As for API it is complicated, as the `flatten` API function will/may be called with encryption already loaded. May be we may add a restriction to the API `flatten` function, e.g. that it should be called with encryption not loaded (then it would format the images as the first step and then load the encryption before proceeding) or that it should be called only for already formatted clones? I would prefer the second.

Reply:
Sounds fair to me.
However I would prefer that the header copyup API won't be exposed by `rbd_encryption_format(...RBD_ENCRYPTION_FORMAT_PARENT,...)`, but rather via a new `rbd_encryption_flatten(<parent_encryption_format>,...)`.

Reply:
> For the record, I must say I don't like it so much, for 2 reasons:

Yeah, my primary concern with it is backwards compatibility (i.e. your first reason).  If it wasn't for that, I think the uniformity that falls out actually has considerable merits, even if some disk space gets wasted.  But all in all, I don't like it much either.

> As for API it is complicated, as the `flatten` API function will/may be called with encryption already loaded. May be we may add a restriction to the API `flatten` function, e.g. that it should be called with encryption not loaded (then it would format the images as the first step and then load the encryption before proceeding)

We would need a new `rbd_flatten` variant, taking the same parameters as `rbd_encryption_load`, in that case.  If one starts with encryption not loaded, they need to get the passphrases from somewhere.  And that would make proxying flattens basically impossible -- I don't see a way to pass passphrases securely in an async request RPC.

> or that it should be called only for already formatted clones? I would prefer the second.

This still heavily special cases unformatted clones at the API surface, just without disk space wastage.

Reply:
> But anyhow, I don't see this debate as critical. So as far as I'm concerned I'm planning to go-on next week with implementing along the lines suggested above.

@orozery Let me take a couple of days to take a deeper look at the code and see if I can come up with an approach to keep the nice `rbd_flatten` "just works" semantics that is more compact than https://github.com/ceph/ceph/pull/47601.  In the meantime, it would be good to address all other outstanding comments (without changing the base commit).

Reply:
I think I already addressed the review you posted a few weeks ago.
There are a few other nits you posted a few days ago which should be addressed in the changes I pushed today.

Reply:
> I think I already addressed the review you posted a few weeks ago. There are a few other nits you posted a few days ago which should be addressed in the changes I pushed today.

Those are indeed addressed but there are still three nits (one in `luks/test_mock_FlattenRequest.cc`, one in `luks/test_mock_LoadRequest.cc` and one in `luks/LoadRequest.cc`) from July:

- https://github.com/ceph/ceph/pull/40363#discussion_r920400179
- https://github.com/ceph/ceph/pull/40363#discussion_r922851775
- https://github.com/ceph/ceph/pull/40363#discussion_r922852377

As for this flattening issue, I'm still experimenting.  I believe I have something promising but it is not complete (yet to be fully compiled even -- I keep being distracted).  I will update this thread when I have more.

Reply:
I finally got some time and made it to work with a very small patch (~250 lines) by defining
```
enum class ImageArea {
  DATA,
  CRYPTO_HEADER
};

struct ImageExtents {
  Extents extents;
  ImageArea area = ImageArea::DATA;

  ...
};
```
and making heavy use of implicit converting constructors to convert from "bare" `Extents` (which are assumed to belong to the `DATA` area) to `ImageExtents` and back.  However, the more I think about it the less I like the implicitness -- it seems worthwhile to require explicit marking of all places that assume the `DATA` area to avoid confusion and nasty cache aliasing bugs in the future.  That would result in a larger patch but it still seems more on point than just passing a bool or a flag (depending on whether the method in question already takes some flags) through everything because what we are really dealing with two separate partitions/areas of the image, one of which is direct-mapped and the other is at an offset and `ImageArea` concept captures that much better.  I'm currently backing out those converting constructors one at time, let's see how it goes.

While testing my flattening patch, I also discovered an I/O path bug which is present in this PR as is.  The cases of a parent with a crypto header of size X and a clone with a crypto header of size Y, where X > Y, aren't handled correctly.  E.g. if a parent is formatted as LUKS2 and a clone is formatted as LUKS1, any write to the last 12M of the clone goes bad because it generates a copyup request that is beyond the size of the parent.  This seems to be down to the parent overlap calculation logic.  I have a tentative fix but it's not validated yet.

Reply:
> I'm currently backing out those converting constructors one at time, let's see how it goes.

In the end I got rid of `ImageExtents` entirely.  I pushed what I have so far on top of this PR to https://github.com/idryomov/ceph/commits/wip-crypto-flatten, please take a look!  This should handle all `ImageDispatcher` operations, not just read and write.  Note that unit tests aren't converted yet -- I got fed up with fixing them after each step and will do it once we agree on the final revision.

> While testing my flattening patch, I also discovered an I/O path bug which is present in this PR as is.

I'm working on the fix right now.

Reply:
LGTM

Reply:
LGTM.

Two suggestions though:

While we're changing `remap_extents` in a way that further point out that `CryptoImageDispatchLayer` is not a real dispatch layer,
how about removing it all-together? (PR #43804)
This is also a requirement for the bug fix in PR #44366.

I would also squash the "revert" type of commits (`librbd: don't temporarily shut down crypto when flattening `,
 `librbd: no need to return format from ShutDownCryptoRequest`) into older commits of my PR.

Reply:
> While we're changing `remap_extents` in a way that further point out that `CryptoImageDispatchLayer` is not a real dispatch layer, how about removing it all-together? (PR #43804) This is also a requirement for the bug fix in PR #44366.

I think #43804 can be rebased on top the of new `remap_to_physical` and `remap_to_logical` methods after this is merged.  I didn't want to complicate these changes by bundling something that isn't strictly necessary (and breaking more unit tests, etc).

> 
> I would also squash the "revert" type of commits (`librbd: don't temporarily shut down crypto when flattening `, `librbd: no need to return format from ShutDownCryptoRequest`) into older commits of my PR.

Older commits would have needed to be rebased on top of some of the prerequisite commits (such as the introduction of `ImageArea` in the first place) and I wanted to avoid that to make these new changes easier to review.

Reply:
On a different matter: what about handling of a flatten where the remote client does not have encryption loaded?
Should we add a new `EncryptedFlattenPayload` as @trociny suggested (which means also adding `rbd_encryption_flatten` instead of using `rbd_flatten`)? Or should we change the existing flatten payload to encode a bit indicating whether the requester client has encryption loaded?

Reply:
> On a different matter: what about handling of a flatten where the remote client does not have encryption loaded? Should we add a new `EncryptedFlattenPayload` as @trociny suggested (which means also adding `rbd_encryption_flatten` instead of using `rbd_flatten`)?

Do you mean adding `rbd_encryption_flatten` method to librbd API?  I'm not sure I understand the need for that: couldn't today's `rbd_flatten` issue either `FlattenPayload` or `EncryptedFlattenPayload` RPC depending on whether encryption is loaded?

> Or should we change the existing flatten payload to encode a bit indicating whether the requester client has encryption loaded?

Conceptually, this is much better but I don't see a good way to make that work given how these payloads are encoded.  New, optional for older clients, fields can be easily added but the semantics of this bit are different -- it's not optional.  An older client that doesn't know anything about encryption has to deny encrypted flatten request (i.e. treat it as an unknown op).

Reply:
Thinking a bit more about it, I'm not sure we need to go out of our way to handle this case at all.  If a client doesn't have encryption loaded but is actively doing something with the image, even if it properly denies encrypted flatten request from the peer, it would most likely break later because its higher level caches would get out of sync (as encrypted flatten done by the peer would change the contents the image from that client's POV).  The same can happen with regular writes as well and there is no check we can add for that case.

So maybe just leave it as is?

Reply:
> Do you mean adding `rbd_encryption_flatten` method to librbd API? I'm not sure I understand the need for that: couldn't today's `rbd_flatten` issue either `FlattenPayload` or `EncryptedFlattenPayload` RPC depending on whether encryption is loaded?

You're right, I didn't think this through :)




> Thinking a bit more about it, I'm not sure we need to go out of our way to handle this case at all. If a client doesn't have encryption loaded but is actively doing something with the image, even if it properly denies encrypted flatten request from the peer, it would most likely break later because its higher level caches would get out of sync (as encrypted flatten done by the peer would change the contents the image from that client's POV). The same can happen with regular writes as well and there is no check we can add for that case.
> 
> So maybe just leave it as is?

Well there's 2 legitimate cases I can think of:
1. A client which only does reads on the raw image (such as exporting). I guess in this case is fine since exclusive lock will only be held by a writing client?
2. A client that intends to write, but still didn't run `rbd_encryption_load`. I guess in that case, again, the exclusive lock won't be acquired by the client since he made no writes (he only did `rbd_open`)?

So assuming I missed no other legit case, I think I agree we can leave it as is.
@dannyharnik did I miss anything?

Reply:
> > While testing my flattening patch, I also discovered an I/O path bug which is present in this PR as is.
> 
> I'm working on the fix right now.

The I/O path issues are now resolved but I ran into another bug.  The crux of it is that shrinking the image boils down to basically the same thing as flattening the affected range in case snapshots are involved.  In addition, even if a clone doesn't have any snapshots, shrinking it to a size that isn't an object boundary is going to trigger a copyup for that object.  This means that `rbd_resize` needs to respect the loaded encryption just as `rbd_flatten` lest data corruption occurs...

I have started working on the fix -- since this resize issue is again brought by this PR as is I'm considering it a merge blocker.

Reply:
I think all blockers are addressed now.  Because the additional diff is pretty large, I'm leaning towards merging this PR as is and opening a new "part 2" PR.  It should be much easier to review that way.  Any objections?

Reply:
Sounds good to me

Reply:
Great, I'm cleaning up fixup commits right now -- should be able to push later today!